viewer-prototype/src/browser/trace-viewer/components/data-providers/tsp-data-provider.ts

```diff
@@ -100,35 +99,34 @@ export class TspDataProvider {
     protected getStateModelByRow(row: TimeGraphRow, chartStart: number): TimelineChart.TimeGraphRowElementModel[] {
         const states: TimelineChart.TimeGraphRowElementModel[] = [];
         row.states.forEach((state: TimeGraphState, idx: number) => {
-            const end = state.startTime + state.duration - chartStart;
-            if (state.style || state.value >= 0) {
+            const end = state.end - chartStart;
+            if (state.style) {
                 states.push({
-                    id: (row as any).entryID + '-' + idx,
-                    label: state.label,
+                    id: row.entryId + '-' + idx,
+                    label: state.label || '',
                     range: {
-                        start: state.startTime - chartStart,
+                        start: state.start - chartStart,
                         end
                     },
                     data: {
-                        stateValue: state.value,
                         style: state.style
                     }
                 });
                 this.totalRange = this.totalRange < end ? end : this.totalRange;
             } else {
                 const nextIndex = idx + 1;
                 const nextState = row.states[nextIndex];
-                if (nextState && nextState.value < 0) {
+                if (nextState && nextState.start > state.end + 1) {
                     // Add gap state
                     states.push({
-                        id: (row as any).entryID + '-' + idx,
+                        id: row.entryId + '-' + idx,
                         label: 'GAP',
                         range: {
                             start: end,
-                            end: nextState.startTime - chartStart
+                            end: nextState.start - chartStart
                         },
                         data: {
-                            stateValue: -1
+
```

Comment:
could you comment on why a null state isn't -1?

Reply:
the concept of stateValue is now gone from the TSP!!... 

though as I did this patch, I'm ambivalent... stateValue was small number, the style is more verbose. But having both style and value was confusing, as it's either one or the other. And style offers more possibility. 

For now in the Trace Server, when the style is not set in the state, it's using the value's string value as a style... And if it's a null state, then there's no style at all.

Reply:
I think we could fix the style vs value using mixed type in the protocol, as described here: https://swagger.io/docs/specification/data-models/data-types/